src/tof.py

```diff
@@ -0,0 +1,148 @@
+# Usage:
+
+# cd esp-wheelchair
+# python3 software/tof.py
+
+import rospy
+from std_msgs.msg import Float64
+from std_msgs.msg import Int16
+from sensor_msgs.msg import PointCloud2
+from geometry_msgs.msg import Twist
+import pcl
+import ros_numpy
+import numpy as np
+import pcl.pcl_visualization
+from manual_control import * 
+from repelent_field_control import *
+
+
+
+# Parameters
+useVisual = True
+PWM_MIN = 5
+PWMRANGE = 40
+MODE = None
+# variable initialization
+emergencyStopThreshold = 0.1
+if(useVisual):
+    viewer_front = pcl.pcl_visualization.PCLVisualizering()
+    viewer_back = pcl.pcl_visualization.PCLVisualizering()
+inputLinear = None
+inputAngular = None
+manualMode = ManualMode()
+repelentMode = RepelentMode()
+
+sign = lambda a: (a>0) - (a<0)
+
+def mapPwm(x, out_min, out_max):
+    """Map the x value 0.0 - 1.0 to out_min to out_max"""
+    return x * (out_max - out_min) + out_min;
+
+def pointCloud_to_NumpyArray(point_Cloud):
+    """ convert a ROS's pointcloud data structure to a numpy array"""
+    height = point_Cloud.shape[0]
+    width = point_Cloud.shape[1]
+    np_points = np.zeros((height * width, 3), dtype=np.float32)
+    np_points[:, 0] = np.resize(point_Cloud['x'], height * width)
+    np_points[:, 1] = np.resize(point_Cloud['y'], height * width)
+    np_points[:, 2] = np.resize(point_Cloud['z'], height * width)
+    return np_points
+
+def visualizePointCloud(viewer ,point_Cloud):
+    """ visualize a numpy point cloud to a viewer """
+    p = pcl.PointCloud(np.array(point_Cloud, dtype=np.float32))
+    viewer.AddPointCloud(p, b'scene_cloud_front', 0)
+    viewer.SpinOnce()
+    viewer.RemovePointCloud( b'scene_cloud_front', 0)
+
+
+def point_cloud_front_callback(msg):
+    """ Callback function for front ToF sensor """
+
+    # change from pointcloud2 to numpy
+    pc = ros_numpy.numpify(msg)
+    front_Pointcloud_array = pointCloud_to_NumpyArray(pc)
+    # visualize if useVisual is TRUE 
+    if(useVisual):
+        visualizePointCloud(viewer_front, front_Pointcloud_array)
+    # find the nearest point and store it at repelent Class
+    minDist_front =  np.nanmin(front_Pointcloud_array[:,2])
+    repelentMode.setDistanceFront(minDist_front)
+
+def point_cloud_back_callback(msg):
+    """ Callback function for back ToF sensor """
+
+    # change from pointcloud2 to numpy
+    pc = ros_numpy.numpify(msg)
+    back_Pointcloud_array = pointCloud_to_NumpyArray(pc)
+    # visualize if useVisual is TRUE 
+    if(useVisual): 
+        visualizePointCloud(viewer_back, back_Pointcloud_array)
+    # find the nearest point and store it at repelent Class
+    minDist_back = np.nanmin(back_Pointcloud_array[:, 2])
+    repelentMode.setDistanceBack(minDist_back)
```

Comment:
Suggestion: add an empty line between logical sections, each logical section beginning with a small comment about what that logical section does. Like paragraphs in a text. Also applies to the rest of the file.

Reply:
DOne